lib/promise/group.rb

```diff
@@ -21,22 +20,30 @@ def wait
       end
     end
 
+    def promise_fulfilled(_value, _arg)
+      @remaining -= 1
+      if @remaining.zero?
+        result = @inputs.map { |obj| promise?(obj) ? obj.value : obj }
+        promise.fulfill(result)
+      end
+    end
+
+    def promise_rejected(reason, _arg)
+      promise.reject(reason)
+    end
+
     private
 
     def chain_inputs
-      on_fulfill = method(:on_fulfill)
-      on_reject = promise.public_method(:reject)
       each_promise do |input_promise|
-        input_promise.then(on_fulfill, on_reject)
+        if input_promise.pending?
+          input_promise.subscribe(self, nil, nil)
+        elsif input_promise.rejected?
+          return promise_rejected(input_promise.reason, nil)
```

Comment:
Ideally we could get mutant to ignore a return, since removing the `return` causes no behaviour changes, but it doesn't make sense to keep looping after the result promise is rejected.  Any idea how to do that?  Or should we just remove the `return` since this isn't a hot path and will make CI happy?